packages/react-instantsearch-core/src/connectors/connectConfigureRelatedItems.ts

```diff
@@ -0,0 +1,220 @@
+import algoliasearchHelper, {
+  PlainSearchParameters,
+  SearchParameters,
+} from 'algoliasearch-helper';
+import createConnector, { ConnectedProps } from '../core/createConnector';
+import {
+  omit,
+  getObjectType,
+  removeEmptyKey,
+  removeEmptyArraysFromObject,
+} from '../core/utils';
+import {
+  refineValue,
+  getIndexId,
+  hasMultipleIndices,
+} from '../core/indexUtils';
+
+type Hit = any;
+
+export type MatchingPatterns = {
+  [attribute: string]: {
+    /**
+     * The score of the optional filter.
+     *
+     * @see https://www.algolia.com/doc/guides/managing-results/rules/merchandising-and-promoting/in-depth/optional-filters/
+     */
+    score: number;
+  };
+};
+
+interface ConfigureRelatedItemsProps {
+  /**
+   * The reference hit to extract the filters from.
+   */
+  hit: Hit;
+  /**
+   * The schema to create the optional filters.
+   * Each key represents an attribute from the hit.
+   */
+  matchingPatterns: MatchingPatterns;
+  /**
+   * Function to transform the generated search parameters.
+   */
+  transformSearchParameters?(
+    searchParameters: SearchParameters
+  ): PlainSearchParameters;
+}
+
+function createOptionalFilter({
+  attributeName,
+  attributeValue,
+  attributeScore,
+}) {
+  return `${attributeName}:${attributeValue}<score=${attributeScore || 1}>`;
+}
+
+const defaultProps: Partial<ConfigureRelatedItemsProps> = {
+  transformSearchParameters: x => ({ ...x }),
+};
+
+function getId(): string {
+  // We store the search state of this widget in `configure`.
+  return 'configure';
+}
+
+type InternalConfigureRelatedItemsProps = ConfigureRelatedItemsProps &
+  Required<typeof defaultProps>;
+
+function getSearchParametersFromProps(
+  props: ConnectedProps<InternalConfigureRelatedItemsProps>
+): PlainSearchParameters {
+  const optionalFilters = Object.keys(props.matchingPatterns).reduce<
+    Array<string | string[]>
+  >((acc, attributeName) => {
+    const attributePattern = props.matchingPatterns[attributeName];
+    const attributeValue = props.hit[attributeName];
+    const attributeScore = attributePattern.score;
+
+    if (Array.isArray(attributeValue)) {
+      return [
+        ...acc,
+        attributeValue.map(attributeSubValue => {
```

Comment:
I've got a question.
When optionalFilters has a nested array, does it mean the values are used with 'OR' operator?

![image](https://user-images.githubusercontent.com/499898/70806149-f3d27500-1dba-11ea-9f2a-4290e5ccb79d.png)

For example, that's equal to the following?

> attribute=value AND (attribute1=value OR attribute2=value)

I couldn't find the explanation from [here](https://www.algolia.com/doc/api-reference/api-parameters/optionalFilters/?language=javascript).

Reply:
When we use the array syntax, the same rules as `facetFilters` apply. You can read the [usage notes](https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/#usage-notes).

In your example, it would get translated to:

```
attribute:value AND (attribute1:value OR attribute2:value)
```

We're manipulating optional filters here, so whenever something matches, it gets ranked higher.

Reply:
Thanks for the link! I thought I saw it somewhere but couldn't find it 👍